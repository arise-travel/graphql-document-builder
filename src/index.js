// @flow

import * as t from 'graphql-ast-types'
import {
  isFunction,
  isString,
  isInteger,
  isNumber,
  isBoolean,
  isNil,
  isArray,
  isPlainObject,
  toPairs
} from 'lodash'
import {
  assertIsOfTypeOrNull,
  assertValidString,
  assertValidStringOrNull,
  assertValidOperation
} from './validation'
import { print } from 'graphql/language'

// GQL Helpers

export function jsToGQLValue (val) {
  if (t.isValue(val)) {
    return val
  }

  if (isBoolean(val)) {
    return t.booleanValue(val)
  }

  if (isNil(val)) {
    return t.nullValue()
  }

  if (isString(val)) {
    return t.stringValue(val)
  }

  if (isInteger(val)) {
    return t.intValue(val + '') // Ridiculous, but needed
  }

  if (isNumber(val)) {
    return t.floatValue(val + '') // Here too
  }

  if (isArray(val)) {
    return t.listValue(val.map(v => jsToGQLValue(v)))
  }

  if (isPlainObject(val)) {
    return t.objectValue(
      toPairs(val).map(pair =>
        t.objectField(t.name(pair[0]), jsToGQLValue(pair[1]))
      )
    )
  }

  throw new Error(
    `Could not guess GraphQL type of value ${val}. Please use types from 'graphql-ast-types', as 't.enumValue()' for example`
  )
}

// Builders

export class ArgumentList {
  constructor () {
    this.args = []
  }

  addArgument (name, value) {
    assertValidString(name, 'Argument name')
    const gqlValue = jsToGQLValue(value)
    const arg = { name, value: gqlValue }
    this.args.push(arg)
    return this
  }

  toASTNodeArray () {
    return this.args.map(({ name, value }) => t.argument(t.name(name), value))
  }

  static build (args) {
    let argumentList = new ArgumentList()
    if (isFunction(args)) {
      args(argumentList)
    } else if (isPlainObject(args)) {
      toPairs(args).forEach(([name, value]) => {
        argumentList.addArgument(name, value)
      })
    } else {
      assertIsOfTypeOrNull(args, ArgumentList)
      argumentList = args
    }

    return argumentList
  }
}

export class Field {
  constructor () {
    this.name = null
    this.alias = null
    this.argList = null
    this.selectionSet = null
  }

  setName (name) {
    assertValidString(name, 'Field name')
    this.name = name
    return this
  }

  setAlias (alias) {
    assertValidStringOrNull(alias, 'Field alias')
    this.alias = alias
    return this
  }

  /**
   * @param ArgumentList|Function<ArgumentList> args
   */
  setArguments (args = null) {
    if (args === null) {
      this.argList = null
      return this
    }

    this.argList = ArgumentList.build(args)
    return this
  }

  setSelectionSet (args = null) {
    if (args === null) {
      this.selectionSet = null
      return this
    }

    this.selectionSet = SelectionSet.build(args)
    return this
  }

  toASTNode () {
    return t.field(
      t.name(this.name),
      this.alias ? t.name(this.alias) : null,
      this.argList ? this.argList.toASTNodeArray() : null,
      null, // Ignoring directives
      this.selectionSet ? this.selectionSet.toASTNode() : null
    )
  }

  print () {
    return print(this.toASTNode())
  }

  // TODO : switch to single argument, for better semantics across builders
  static build (name, alias = null, args = null, selectionSet = null) {
    if (name instanceof Field) {
      return name
    }

    const field = new Field()
    field
      .setName(name)
      .setAlias(alias)
      .setArguments(args)
      .setSelectionSet(selectionSet)

    return field
  }
}

export class SelectionSet {
  constructor () {
    this.fields = []
  }

  addField (...args) {
    const field = Field.build(...args)
    this.fields.push(field)
    return this
  }

  toASTNode () {
    return t.selectionSet(
      this.fields.map(field => field.toASTNode())
    )
  }

  print () {
    return print(this.toASTNode())
  }

  static build (args) {
    if (args instanceof SelectionSet) return args

    const selectionSet = new SelectionSet()
    if (isArray(args)) {
      args.forEach(arg => {
        if (isArray(arg)) selectionSet.addField(...arg)
        else selectionSet.addField(arg)
      })
    } else if (isFunction(args)) {
      args(selectionSet)
    } else {
      throw new Error('Cannot build SelectionSet with given arguments')
    }

    return selectionSet
  }
}

export class OperationDefinition {
  constructor () {
    this.name = null
    this.selectionSet = new SelectionSet()
    this.operation = 'query'
  }

  addField (...args) {
    this.selectionSet.addField(...args)
    return this
  }

  setName (name) {
    assertValidStringOrNull(name, 'Operation name')
    this.name = name
    return this
  }

  setOperation (operation) {
    assertValidOperation(operation)
    this.operation = operation
    return this
  }

  setSelectionSet (...args) {
    this.selectionSet = SelectionSet.build(...args)
    return this
  }

  toASTNode () {
    return t.operationDefinition(
      this.operation,
      this.selectionSet.toASTNode(),
      this.name ? t.name(this.name) : null
    )
  }

  print () {
    return print(this.toASTNode())
  }

  /**
   * @param {string} type The type of the operation
   */
  static build (type, args) {
    if (type instanceof OperationDefinition) return type

    const operation = new OperationDefinition()
    let realArgs = type
    if (isString(type)) {
      operation.setOperation(type)
      realArgs = args
      if (isArray(args)) {
        realArgs = [type, ...args]
      }
    }
    if (isArray(realArgs)) {
      operation.setOperation(realArgs[0])
      operation.setSelectionSet(realArgs[1])
      operation.setName(realArgs[2])
    } else if (isFunction(realArgs)) {
      realArgs(operation)
    } else {
      throw new Error('Cannot build OperationDefinition with given arguments')
    }

    return operation
  }
}

export class Query extends OperationDefinition {
  static build (args) {
    return OperationDefinition.build('query', args)
  }
}

export class Mutation extends OperationDefinition {
  constructor () {
    super()
    this.operation = 'mutation'
  }

  static build (args) {
    return OperationDefinition.build('mutation', args)
  }
}

export class Subscription extends OperationDefinition {
  constructor () {
    super()
    this.operation = 'subscription'
  }

  static build (args) {
    return OperationDefinition.build('subscription', args)
  }
}

export class Document {
  constructor () {
    this.operations = []
  }

  addQuery (...args) {
    return this.addOperation('query', ...args)
  }

  addMutation (...args) {
    return this.addOperation('mutation', ...args)
  }

  addSubscription (...args) {
    return this.addOperation('subscription', ...args)
  }

  addOperation (type, ...args) {
    if (type instanceof OperationDefinition) {
      this.operations.push(type)
      return this
    }

    const operation = OperationDefinition.build(type, args)
    this.operations.push(operation)
    return this
  }

  static build (args) {
    const document = new Document()

    if (isArray(args)) {
      args.forEach(a => document.addOperation(a))
    } else if (isFunction(args)) {
      args(document)
    } else if (isPlainObject(args)) {
      ['query', 'mutation', 'subscription'].forEach((type) => {
        if (args[type]) {
          document.addOperation(type, args[type], args.name)
        }
      })
    } else {
      throw new Error('Cannot build Document with given arguments')
    }

    return document
  }

  toASTNode () {
    return t.document(this.operations.map(op => op.toASTNode()))
  }

  print () {
    return print(this.toASTNode())
  }
}
