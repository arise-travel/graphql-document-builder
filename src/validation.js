import { trim, isNil } from 'lodash'

function _orNullDecorator (fn) {
  return function (arg, ...rest) {
    if (isNil(arg)) return
    return fn(arg, ...rest)
  }
}

export function assertValidString (str, name) {
  const trimmed = trim(str)
  if (trimmed === '') {
    throw new Error(`${name} should not be empty`)
  }
}

export const assertValidStringOrNull = _orNullDecorator(assertValidString)

export function assertIsOfType (obj, type) {
  if (!(obj instanceof type)) {
    throw new Error(`${obj} is not an instance of ${type}`)
  }
}

export const assertIsOfTypeOrNull = _orNullDecorator(assertIsOfType)

export function assertValidOperation (operation) {
  if (['query', 'mutation', 'subscription'].indexOf(operation) === -1) {
    throw new Error(`${operation} is not a valid operation (query, mutation or subscription)`)
  }
}
