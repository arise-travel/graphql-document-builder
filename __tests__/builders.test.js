import { ArgumentList, Field, SelectionSet, OperationDefinition, Document, Query, Mutation, Subscription } from '../src'
import * as t from 'graphql-ast-types'

describe('ArgumentList builder', () => {
  test('as function', () => {
    const list = ArgumentList.build((list) => {
      expect(list).toBeInstanceOf(ArgumentList)
      list.addArgument('test', 'value')
      list.addArgument('test2', 5)
    })

    expect(list.toASTNodeArray()).toMatchSnapshot()
  })

  test('as object', () => {
    const list = ArgumentList.build({
      test: 'value',
      test2: 5
    })

    expect(list.toASTNodeArray()).toMatchSnapshot()
  })
})

describe('Field builder', () => {
  test('builder', () => {
    let field = Field.build('name').toASTNode()
    expect(field.name.value).toBe('name')

    field = Field.build('name', 'alias').toASTNode()
    expect(field.alias.value).toBe('alias')

    field = Field.build('name', null, { arg: 'list' }).toASTNode()
    field.arguments.forEach(arg => expect(t.isArgument(arg)).toBe(true))

    field = Field.build('name', null, null, ['field1', 'field2']).toASTNode()
    expect(t.isSelectionSet(field.selectionSet)).toBe(true)
  })
})

describe('SelectionSet builder', () => {
  test('as array', () => {
    const selectionSet = SelectionSet.build(['field', ['name', 'alias', { args: true }, ['nested', ['fields']]]])
    expect(selectionSet.toASTNode()).toMatchSnapshot()
  })

  test('as function', () => {
    const selectionSet = SelectionSet.build((set) => {
      expect(set).toBeInstanceOf(SelectionSet)
      set.addField('name')
      set.addField('other', 'alias')
      set.addField('with', null, { args: 'too' })
      set.addField('nested', null, null, ['fieldsAsArray'])
      set.addField('or', null, null, (set) => {
        expect(set).toBeInstanceOf(SelectionSet)
        set.addField('asFunction')
      })
    }).toASTNode()

    expect(selectionSet).toMatchSnapshot()
  })
})

describe('OperationDefinition builders', () => {
  test('as array', () => {
    const operation = OperationDefinition.build([
      'mutation',
      [
        'field1',
        ['field2', 'alias']
      ],
      'mutationName'
    ]).toASTNode()

    expect(t.isOperationDefinition(operation)).toBe(true)
    expect(operation.name.value).toBe('mutationName')
    expect(t.isSelectionSet(operation.selectionSet)).toBe(true)
    expect(operation.operation).toBe('mutation')
  })

  test('as function', () => {
    const operation = OperationDefinition.build((op) => {
      expect(op).toBeInstanceOf(OperationDefinition)
      op
        .setName('mutationName')
        .setOperation('mutation')
        .addField('field1')
        .addField('field2', null, { args: 1 })
    }).toASTNode()

    expect(t.isOperationDefinition(operation)).toBe(true)
    expect(operation.name.value).toBe('mutationName')
    expect(t.isSelectionSet(operation.selectionSet)).toBe(true)
    expect(operation.operation).toBe('mutation')
  })

  test('with type prefix', () => {
    {
      const operation = OperationDefinition.build('mutation', [
        [
          'field1',
          ['field2', 'alias']
        ],
        'mutationName'
      ]).toASTNode()

      expect(t.isOperationDefinition(operation)).toBe(true)
      expect(operation.operation).toBe('mutation')
    }
    {
      const operation = OperationDefinition.build('subscription', (op) => {
        expect(op).toBeInstanceOf(OperationDefinition)
        op
          .addField('field1')
          .addField('field2', null, { args: 1 })
      }).toASTNode()

      expect(t.isOperationDefinition(operation)).toBe(true)
      expect(operation.operation).toBe('subscription')
    }
  })

  test('operation aliases', () => {
    {
      const operation = Mutation.build([
        [
          'field1',
          ['field2', 'alias']
        ],
        'mutationName'
      ]).toASTNode()

      expect(t.isOperationDefinition(operation)).toBe(true)
      expect(operation.operation).toBe('mutation')
    }
    {
      const operation = Subscription.build((op) => {
        expect(op).toBeInstanceOf(OperationDefinition)
        op
          .addField('field1')
          .addField('field2', null, { args: 1 })
      }).toASTNode()

      expect(t.isOperationDefinition(operation)).toBe(true)
      expect(operation.operation).toBe('subscription')
    }
    {
      const operation = Query.build((op) => {
        expect(op).toBeInstanceOf(OperationDefinition)
        op
          .addField('field1')
          .addField('field2', null, { args: 1 })
          .setName('named')
      }).toASTNode()

      expect(t.isOperationDefinition(operation)).toBe(true)
      expect(operation.operation).toBe('query')
      expect(operation.name.value).toBe('named')
    }
  })
})

describe('Document builder', () => {
  test('as array', () => {
    const document = Document.build([
      [
        'mutation',
        [
          'field1',
          ['field2', 'alias']
        ]
      ],
      [
        'query',
        (query) => {
          query.addField('test')
        },
        'named'
      ]
    ]).toASTNode()

    expect(t.isDocument(document)).toBe(true)
    expect(document.definitions[0].operation).toBe('mutation')
    expect(document.definitions[0].selectionSet.selections[0].name.value).toBe('field1')
    expect(document.definitions[0].selectionSet.selections[1].name.value).toBe('field2')
    expect(document.definitions[0].selectionSet.selections[1].alias.value).toBe('alias')
    expect(document.definitions[1].operation).toBe('query')
    expect(document.definitions[1].name.value).toBe('named')
    expect(document.definitions[1].selectionSet.selections[0].name.value).toBe('test')
  })

  test('as function', () => {
    const document = Document.build((d) => {
      d
        .addMutation([
          'field1',
          ['field2', 'alias']
        ])
        .addQuery(
          (query) => {
            query.addField('test')
          },
          'named'
        )
    }).toASTNode()

    expect(t.isDocument(document)).toBe(true)
    expect(document.definitions[0].operation).toBe('mutation')
    expect(document.definitions[0].selectionSet.selections[0].name.value).toBe('field1')
    expect(document.definitions[0].selectionSet.selections[1].name.value).toBe('field2')
    expect(document.definitions[0].selectionSet.selections[1].alias.value).toBe('alias')
    expect(document.definitions[1].operation).toBe('query')
    expect(document.definitions[1].name.value).toBe('named')
    expect(document.definitions[1].selectionSet.selections[0].name.value).toBe('test')
  })

  test('as object shortcut', () => {
    {
      // Testing shorthand for single operation doucment
      const document = Document.build({
        mutation: [
          'field1',
          ['field2', 'alias']
        ]
      }).toASTNode()
      expect(t.isDocument(document)).toBe(true)
      expect(document.definitions[0].operation).toBe('mutation')
      expect(document.definitions[0].selectionSet.selections[0].name.value).toBe('field1')
      expect(document.definitions[0].selectionSet.selections[1].name.value).toBe('field2')
      expect(document.definitions[0].selectionSet.selections[1].alias.value).toBe('alias')
    }
    {
      // Testing shorthand for single operation doucment
      const document = Document.build({
        name: 'named',
        query: (q) => {
          q.addField('test')
        }
      }).toASTNode()

      expect(t.isDocument(document)).toBe(true)
      expect(document.definitions[0].operation).toBe('query')
      expect(document.definitions[0].name.value).toBe('named')
      expect(document.definitions[0].selectionSet.selections[0].name.value).toBe('test')
    }
  })
})
