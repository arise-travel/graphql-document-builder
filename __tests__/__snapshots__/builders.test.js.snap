// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArgumentList builder as function 1`] = `
Array [
  Object {
    "kind": "Argument",
    "name": Object {
      "kind": "Name",
      "value": "test",
    },
    "value": Object {
      "kind": "StringValue",
      "value": "value",
    },
  },
  Object {
    "kind": "Argument",
    "name": Object {
      "kind": "Name",
      "value": "test2",
    },
    "value": Object {
      "kind": "IntValue",
      "value": "5",
    },
  },
]
`;

exports[`ArgumentList builder as object 1`] = `
Array [
  Object {
    "kind": "Argument",
    "name": Object {
      "kind": "Name",
      "value": "test",
    },
    "value": Object {
      "kind": "StringValue",
      "value": "value",
    },
  },
  Object {
    "kind": "Argument",
    "name": Object {
      "kind": "Name",
      "value": "test2",
    },
    "value": Object {
      "kind": "IntValue",
      "value": "5",
    },
  },
]
`;

exports[`SelectionSet builder as array 1`] = `
Object {
  "kind": "SelectionSet",
  "selections": Array [
    Object {
      "alias": null,
      "arguments": null,
      "directives": null,
      "kind": "Field",
      "name": Object {
        "kind": "Name",
        "value": "field",
      },
      "selectionSet": null,
    },
    Object {
      "alias": Object {
        "kind": "Name",
        "value": "alias",
      },
      "arguments": Array [
        Object {
          "kind": "Argument",
          "name": Object {
            "kind": "Name",
            "value": "args",
          },
          "value": Object {
            "kind": "BooleanValue",
            "value": true,
          },
        },
      ],
      "directives": null,
      "kind": "Field",
      "name": Object {
        "kind": "Name",
        "value": "name",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": null,
            "arguments": null,
            "directives": null,
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "nested",
            },
            "selectionSet": null,
          },
          Object {
            "alias": null,
            "arguments": null,
            "directives": null,
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "fields",
            },
            "selectionSet": null,
          },
        ],
      },
    },
  ],
}
`;

exports[`SelectionSet builder as function 1`] = `
Object {
  "kind": "SelectionSet",
  "selections": Array [
    Object {
      "alias": null,
      "arguments": null,
      "directives": null,
      "kind": "Field",
      "name": Object {
        "kind": "Name",
        "value": "name",
      },
      "selectionSet": null,
    },
    Object {
      "alias": Object {
        "kind": "Name",
        "value": "alias",
      },
      "arguments": null,
      "directives": null,
      "kind": "Field",
      "name": Object {
        "kind": "Name",
        "value": "other",
      },
      "selectionSet": null,
    },
    Object {
      "alias": null,
      "arguments": Array [
        Object {
          "kind": "Argument",
          "name": Object {
            "kind": "Name",
            "value": "args",
          },
          "value": Object {
            "kind": "StringValue",
            "value": "too",
          },
        },
      ],
      "directives": null,
      "kind": "Field",
      "name": Object {
        "kind": "Name",
        "value": "with",
      },
      "selectionSet": null,
    },
    Object {
      "alias": null,
      "arguments": null,
      "directives": null,
      "kind": "Field",
      "name": Object {
        "kind": "Name",
        "value": "nested",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": null,
            "arguments": null,
            "directives": null,
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "fieldsAsArray",
            },
            "selectionSet": null,
          },
        ],
      },
    },
    Object {
      "alias": null,
      "arguments": null,
      "directives": null,
      "kind": "Field",
      "name": Object {
        "kind": "Name",
        "value": "or",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": null,
            "arguments": null,
            "directives": null,
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "asFunction",
            },
            "selectionSet": null,
          },
        ],
      },
    },
  ],
}
`;
